[Bug] Docker

Usually solved by retrying / de-provisioning and re-provisioning but need a graceful way to deal with it. 

```
 21:09:31 [DEBUG] Install gateway package on instances: 50.90s
Traceback (most recent call last):
  File "/home/ubuntu/miniconda/bin/skylark", line 33, in <module>
    sys.exit(load_entry_point('skylark', 'console_scripts', 'skylark')())
  File "/home/ubuntu/miniconda/lib/python3.8/site-packages/typer/main.py", line 214, in __call__
    return get_command(self)(*args, **kwargs)
  File "/home/ubuntu/miniconda/lib/python3.8/site-packages/click/core.py", line 1128, in __call__
    return self.main(*args, **kwargs)
  File "/home/ubuntu/miniconda/lib/python3.8/site-packages/click/core.py", line 1053, in main
    rv = self.invoke(ctx)
  File "/home/ubuntu/miniconda/lib/python3.8/site-packages/click/core.py", line 1659, in invoke
    return _process_result(sub_ctx.command.invoke(sub_ctx))
  File "/home/ubuntu/miniconda/lib/python3.8/site-packages/click/core.py", line 1395, in invoke
    return ctx.invoke(self.callback, **ctx.params)
  File "/home/ubuntu/miniconda/lib/python3.8/site-packages/click/core.py", line 754, in invoke
    return __callback(*args, **kwargs)
  File "/home/ubuntu/miniconda/lib/python3.8/site-packages/typer/main.py", line 500, in wrapper
    return callback(**use_params)  # type: ignore
  File "/home/ubuntu/skylark/skylark/cli/cli.py", line 245, in replicate_json
    rc.provision_gateways(reuse_gateways)
  File "/home/ubuntu/skylark/skylark/replicate/replicator_client.py", line 187, in provision_gateways
    do_parallel(lambda arg: arg[0].start_gateway(arg[1], gateway_docker_image=self.gateway_docker_image), args, n=-1)
  File "/home/ubuntu/skylark/skylark/utils/utils.py", line 72, in do_parallel
    args, result = future.result()
  File "/home/ubuntu/miniconda/lib/python3.8/concurrent/futures/_base.py", line 437, in result
    return self.__get_result()
  File "/home/ubuntu/miniconda/lib/python3.8/concurrent/futures/_base.py", line 389, in __get_result
    raise self._exception
  File "/home/ubuntu/miniconda/lib/python3.8/concurrent/futures/thread.py", line 57, in run
    result = self.fn(*self.args, **self.kwargs)
  File "/home/ubuntu/skylark/skylark/utils/utils.py", line 65, in wrapped_fn
    return args, func(args)
  File "/home/ubuntu/skylark/skylark/replicate/replicator_client.py", line 187, in <lambda>
    do_parallel(lambda arg: arg[0].start_gateway(arg[1], gateway_docker_image=self.gateway_docker_image), args, n=-1)
  File "/home/ubuntu/skylark/skylark/compute/server.py", line 224, in start_gateway
    assert "Status: Downloaded newer image" in docker_out or "Status: Image is up to date" in docker_out, (docker_out, docker_err)
AssertionError: ('', 'Error response from daemon: Get "https://ghcr.io/v2/": net/http: request canceled while waiting for connection (Client.Timeout exceeded while awaiting headers)\n')
```

--- a/skylark/compute/aws/aws_cloud_provider.py
+++ b/skylark/compute/aws/aws_cloud_provider.py
@@ -247,7 +247,7 @@ class AWSCloudProvider(CloudProvider):
         if name is None:
             name = f"skylark-aws-{str(uuid.uuid4()).replace('-', '')}"
         iam_instance_profile_name = f"{name}_profile"
-        iam = self.auth.get_boto3_client("iam")
+        iam = self.auth.get_boto3_client("iam", region)
         ec2 = self.auth.get_boto3_resource("ec2", region)
         vpc = self.get_vpc(region)
         assert vpc is not None, "No VPC found"

--- a/skylark/compute/server.py
+++ b/skylark/compute/server.py
@@ -1,3 +1,4 @@
+from functools import partial
 import json
 import subprocess
 from enum import Enum, auto
@@ -197,6 +198,13 @@ class Server:
         if not docker_version.startswith("Success"):
             raise RuntimeError(f"Failed to install Docker on {self.region_tag}, {self.public_ip()}: OUT {out}\nERR {err}")
 
+    def pull_docker(self, gateway_docker_image):
+        docker_out, docker_err = self.run_command(f"sudo docker pull {gateway_docker_image}")
+        if "Status: Downloaded newer image" not in docker_out and "Status: Image is up to date" not in docker_out:
+            raise RuntimeError(
+                f"Failed to pull docker image {gateway_docker_image} on {self.region_tag}, {self.public_ip()}: OUT {docker_out}\nERR {docker_err}"
+            )
+
     def start_gateway(
         self,
         outgoing_ports: Dict[str, int],  # maps ip to number of connections along route
@@ -226,8 +234,7 @@ class Server:
 
         # pull docker image and start container
         with Timer(f"{desc_prefix}: Docker pull"):
-            docker_out, docker_err = self.run_command(f"sudo docker pull {gateway_docker_image}")
-            assert "Status: Downloaded newer image" in docker_out or "Status: Image is up to date" in docker_out, (docker_out, docker_err)
+            retry_backoff(partial(self.pull_docker, gateway_docker_image), exception_class=RuntimeError)
         logger.debug(f"{desc_prefix}: Starting gateway container")
         docker_run_flags = f"-d --log-driver=local --log-opt max-file=16 --ipc=host --network=host --ulimit nofile={1024 * 1024}"
         docker_run_flags += " --mount type=tmpfs,dst=/skylark,tmpfs-size=$(($(free -b  | head -n2 | tail -n1 | awk '{print $2}')/2))"

--- a/skylark/utils/utils.py
+++ b/skylark/utils/utils.py
@@ -99,6 +99,9 @@ def retry_backoff(
             if i == max_retries - 1:
                 raise e
             else:
-                logger.warning(f"Retrying function due to {e} (attempt {i + 1}/{max_retries})")
+                # ignore retries due to IAM instance profile propagation
+                if "Invalid IAM Instance Profile name" not in str(e):
+                    fn_name = fn.__name__ if hasattr(fn, "__name__") else "unknown function"
+                    logger.warning(f"Retrying {fn_name} due to: {e} (attempt {i + 1}/{max_retries})")
                 time.sleep(backoff)
                 backoff = min(backoff * 2, max_backoff)

